README.md
# Ronin Web

* http://ronin.rubyforge.org/web/
* http://github.com/postmodern/ronin-web
* http://github.com/postmodern/ronin-web/issues
* http://groups.google.com/group/ronin-ruby
* irc.freenode.net #ronin

## DESCRIPTION:

Ronin Web is a Ruby library for Ronin that provides support for web
scraping and spidering functionality.

Ronin is a Ruby platform for exploit development and security research.
Ronin allows for the rapid development and distribution of code, exploits
or payloads over many common Source-Code-Management (SCM) systems.

### Ruby

Ronin's Ruby environment allows security researchers to leverage Ruby with
ease. The Ruby environment contains a multitude of convenience methods
for working with data in Ruby, a Ruby Object Database, a customized Ruby
Console and an extendable command-line interface.

### Extend

Ronin's more specialized features are provided by additional Ronin
libraries, which users can choose to install. These libraries can allow
one to write and run Exploits and Payloads, scan for PHP vulnerabilities,
perform Google Dorks  or run 3rd party scanners.

### Publish

Ronin allows users to publish and share code, exploits, payloads or other
data via Overlays. Overlays are directories of code and data that can be
hosted on any SVN, Hg, Git or Rsync server. Ronin makes it easy to create,
install or update Overlays.

## FEATURES:

* Web access (utilizing Mechanize and Nokogiri).
* Integrates Spidr into Ronin::Web::Spider.
* Provides Ronin::Web::Server, a customizable Sinatra based Web Server that
  supports path and host-name routing.
* Provides Ronin::Web::Proxy, a configurable Sinatra based Web Proxy.

## SYNOPSIS:

Start the Ronin console with Ronin Web preloaded:

    $ ronin-web

## EXAMPLES:

Get a web-page:

    Web.get('http://www.rubyinside.com/')

Get only the body of the web-page:

    Web.get_body('http://www.rubyinside.com/')

Get a WWW::Mechanize agent:

    agent = Web.agent

Parse HTML:

    Web.html(open('some_file.html'))
    # => <!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
    # <html>
    #   <head>
    #     <script type="text/javascript" src="redirect.js"></script>
    #   </head>
    # </html>

Build a HTML document:

    doc = Web.build_html do
      html {
        head {
          script(:type => 'text/javascript', :src => 'redirect.js')
        }
      }
    end
    
    puts doc.to_html
    # <!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
    # <html><head><script src="redirect.js" type="text/javascript"></script></head></html>

Parse XML:

    Web.xml(some_text)
    # => <?xml version="1.0"?>
    # <users>
    #   <user>
    #     <name>admin</name>
    #     <password>0mni</password>
    #   </user>
    # </users>


Build a XML document:

    doc = Web.build_xml do
      playlist {
        mp3 {
          file { text('02 THE WAIT.mp3') }
          artist { text('Evil Nine') }
          track { text('The Wait feat David Autokratz') }
          duration { text('1000000000') }
        }
      }
    end
    
    puts doc.to_xml
    # <?xml version="1.0"?>
    # <playlist>
    #   <mp3>
    #     <file>02 THE WAIT.mp3</file>
    #     <artist>Evil Nine</artist>
    #     <track>The Wait feat David Autokratz</track>
    #     <duration>1000000000</duration>
    #   </mp3>
    # </playlist>

Spider a web site:

    Web::Spider.host('www.example.com') do |spider|
      spider.every_url do |url|
        # ...
      end

      spider.every_page do |page|
        # ...
      end
    end

Serve files via a Web Server:

    require 'ronin/web/server'

    Web.server do
      file '/opensearch.xml', '/tmp/test.xml'
      directory '/download/', '/tmp/download/'
    end

    Web.server.get '/test' do
      'Test 1 2 1 2'
    end

## REQUIREMENTS:

* [mechanize](http://mechanize.rubyforge.org/) >= 0.9.3
* [spidr](http://spidr.rubyforge.org/) >= 0.2.0
* [sinatra](http://www.sinatrarb.com/) >= 0.9.4
* [ronin](http://ronin.rubyforge.org/) >= 0.3.0

## INSTALL:

    $ sudo gem install ronin-web

## LICENSE:

Ronin Web - A Ruby library for Ronin that provides support for web
scraping and spidering functionality.

Copyright (c) 2006-2009 Hal Brodigan (postmodern.mod3 at gmail.com)

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA


History.txt
=== 0.2.1 / 2009-10-18

* Require spidr >= 0.2.0.
* Added Ronin::Scanners::Web.
* Added more specs for Ronin::Web and Ronin::Scanners::Web.
* Renamed Ronin::Web.proxy to Ronin::Web.proxy_server, to not conflict
  with the original Ronin::Web.proxy method.

=== 0.2.0 / 2009-09-24

* Require ronin >= 0.3.0.
* Require mechanize >= 0.9.3.
* Require sinatra >= 0.9.4.
* Require rspec >= 1.2.8.
* Require test-unit >= 1.2.3.
* Require rack-test >= 0.4.1.
* Require yard >= 0.2.3.5.
* Added Ronin::Web::Server::Base.
* Added Ronin::Web::Server::Files.
* Added Ronin::Web::Server::Hosts.
* Added Ronin::Web::Server::Proxy.
* Added Ronin::Web::Server::Helpers.
* Added Ronin::Web::Server::Helpers::Files.
* Added Ronin::Web::Server::Helpers::Hosts.
* Added Ronin::Web::Server::Helpers::Proxy.
* Added Ronin::Web::Server::App.
* Added Ronin::Web.server.
* Renamed Ronin::Sessions::Web to Ronin::Network::Helpers::Web.
* Moved to YARD based documentation.
* Updated the project summary and 3-point description for Ronin Web.
* Refactored the Ronin::Web::Server to build ontop of Sinatra.
* Refactored Ronin::Network::Helpers::Web.

=== 0.1.3 / 2009-07-02

* Use Hoe >= 2.0.0.
* Require spidr >= 0.1.9.
* Require rack >= 1.0.0.
* Require ronin >= 0.2.4.
* Added Ronin::Web::Fingerprint.
* Added Web.build_html.
* Added Web.build_xml.
* Allow Web.html to accept a block.
* Allow Web.xml to accept a block.
* Allow Server#return_file to accept a content_type option.
* Make sure Ronin::Web::Server is thread safe.
* Blocks passed to Server.new will not be instance_evaled.
* Removed Spider#default_options.
* Removed Server.run.
* Removed Server#config.

=== 0.1.2 / 2009-03-28

* Added Ronin::Web::Proxy.
* Added diagnostic messages to Ronin::Web::Spider.
* Fixed Rack 0.9.1 compatibility bugs in Ronin::Web::Server.
  * Server#response now uses Rack::Response.
  * Ronin::Web::Server now uses a Rack::Request object instead of the
    standard env Hash.
* Updated specs for Ronin::Web::Server.

=== 0.1.1 / 2009-02-23

* Added a git style sub-command (<tt>ronin-web</tt>) which starts the Ronin
  console with <tt>ronin/web</tt> preloaded.
* Require Nokogiri >= 1.2.0.
* Require Ronin >= 0.2.1.
* Updated Ronin::Web::Server:
  * Properly catch load errors when attempting to load Mongrel.
  * Renamed Server#mount to Server#directory.
* Removed the last reference to Hpricot.
* Fixed a bug when loading the Nokogiri extensions with Nokogiri >= 1.2.0.
* Updated README.txt.

=== 0.1.0 / 2009-01-22

* Initial release.
  * Provides Web.html and Web.xml methods, which use Nokogiri::HTML and
    Nokogiri::XML respectively, to parse HTML and XML content.
  * Supports custom Proxy and User-Agent strings, through
    the Web.proxy and Web.user_agent methods respectively.
  * Provides Web.agent, Web.get and Weg.post methods which access
    WWW::Mechanize.
  * Integrates Spidr::Agent into Web::Spider.
  * Provides Web::Server, a customizable Rack web server that supports path
    and host-name routing.
    * Web::Server will use Mongrel, only if Mongrel is installed, otherwise
      WEBrick will be used.



lib/ronin/web/web.rb
#
# Ronin Web - A Ruby library for Ronin that provides support for web
# scraping and spidering functionality.
#
# Copyright (c) 2006-2009 Hal Brodigan (postmodern.mod3 at gmail.com)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#

require 'ronin/network/http'

require 'uri/http'
require 'nokogiri'
require 'mechanize'
require 'open-uri'

module Ronin
  module Web
    #
    # Parses the body of a document into a HTML document object.
    #
    # @param [String, IO] body
    #   The body of the document to parse.
    #
    # @yield [doc]
    #   If a block is given, it will be passed the newly created document
    #   object.
    #
    # @yieldparam [Nokogiri::HTML::Document] doc
    #   The new HTML document object.
    #
    # @return [Nokogiri::HTML::Document] The new HTML document object.
    #
    def Web.html(body,&block)
      doc = Nokogiri::HTML(body)

      block.call(doc) if block
      return doc
    end

    #
    # Creates a new Nokogiri::HTML::Builder.
    #
    # @yield []
    #   The block that will be used to construct the HTML document.
    #
    # @return [Nokogiri::HTML::Builder]
    #   The new HTML builder object.
    #
    # @example
    #   Web.build_html do
    #     html {
    #       body {
    #         div(:style => 'display:none;') {
    #           object(:classid => 'blabla')
    #         }
    #       }
    #     }
    #   end
    #
    def Web.build_html(&block)
      Nokogiri::HTML::Builder.new(&block)
    end

    #
    # Parses the body of a document into a XML document object.
    #
    # @param [String, IO] body
    #   The body of the document to parse.
    #
    # @yield [doc]
    #   If a block is given, it will be passed the newly created document
    #   object.
    #
    # @yieldparam [Nokogiri::XML::Document] doc
    #   The new XML document object.
    #
    # @return [Nokogiri::XML::Document] The new XML document object.
    #
    def Web.xml(body,&block)
      doc = Nokogiri::XML(body)

      block.call(doc) if block
      return doc
    end

    #
    # Creates a new Nokogiri::XML::Builder.
    #
    # @yield []
    #   The block that will be used to construct the XML document.
    #
    # @return [Nokogiri::XML::Builder]
    #   The new XML builder object.
    #
    # @example
    #  Web.build_xml do
    #    post(:id => 2) {
    #      title { text('some example') }
    #      body { text('this is one contrived example.') }
    #    }
    #  end
    #
    def Web.build_xml(&block)
      Nokogiri::XML::Builder.new(&block)
    end

    #
    # Proxy information for Ronin::Web to use.
    #
    # @return [Network::HTTP::Proxy]
    #   The Ronin Web proxy information.
    #
    # @see Ronin::Network::HTTP.proxy
    #
    def Web.proxy
      Network::HTTP.proxy
    end

    #
    # Creates a HTTP URI based on a Hash of proxy information.
    #
    # @param [Network::HTTP::Proxy, Hash, String] proxy_info
    #   The proxy information.
    #
    # @return [URI::HTTP, nil]
    #   The HTTP URI that represents the proxy. If the proxy is diabled,
    #   +nil+ will be returned.
    #
    # @example
    #   Web.proxy_url
    #   # => "http://www.example.com:8080"
    #
    # @example
    #   Web.proxy_url({:host => 'www.example.com', :port => 8081})
    #   # => "http://www.example.com:8081"
    #
    # @example
    #   Web.proxy_url('www.example.com:9000')
    #   # => "http://www.example.com:9000"
    #
    def Web.proxy_url(proxy_info=Web.proxy)
      proxy = if proxy_info.kind_of?(Hash)
                Network::HTTP::Proxy.new(proxy_info)
              else
                Network::HTTP::Proxy.parse(proxy_info)
              end

      return proxy.url
    end

    #
    # @return [Array]
    #   The supported Web User-Agent Aliases.
    #
    def Web.user_agent_aliases
      WWW::Mechanize::AGENT_ALIASES
    end

    #
    # @return [String, nil]
    #   The Ronin Web User-Agent
    #
    def Web.user_agent
      Network::HTTP.user_agent
    end

    #
    # Sets the Ronin Web User-Agent.
    #
    # @param [String] new_agent
    #   The User-Agent string to use.
    #
    # @return [String]
    #   The new User-Agent string.
    #
    def Web.user_agent=(new_agent)
      Network::HTTP.user_agent = new_agent
    end

    #
    # Sets the Ronin Web User-Agent.
    #
    # @param [String] name
    #   The User-Agent alias to use.
    #
    # @return [String]
    #   The new User-Agent string.
    #
    def Web.user_agent_alias=(name)
      Network::HTTP.user_agent = Web.user_agent_aliases[name.to_s]
    end

    #
    # Opens a URL as a temporary file.
    #
    # @param [Hash] options
    #   Additional options.
    #
    # @option options [String] :user_agent_alias
    #   The User-Agent Alias to use.
    #
    # @option options [String] :user_agent
    #   The User-Agent string to use.
    #
    # @option options [Network::HTTP::Proxy, Hash, String] :proxy
    #   (Web.proxy)
    #   Proxy information.
    #
    # @option options [String] :user
    #   The HTTP Basic Authentication user name.
    #
    # @option options [String] :password
    #   The HTTP Basic Authentication password.
    #
    # @option options [Proc] :content_length_proc
    #   A callback which will be passed the content-length of the HTTP
    #   response.
    #
    # @option options [Proc] :progress_proc
    #   A callback which will be passed the size of each fragment, once
    #   received from the server.
    #
    # @return [File]
    #   The contents of the URL.
    #
    # @example Open a given URL.
    #   Web.open('http://rubyflow.com/')
    #
    # @example Open a given URL, using a custom User-Agent alias.
    #   Web.open('http://tenderlovemaking.com/',
    #     :user_agent_alias => 'Linux Mozilla')
    #
    # @example Open a given URL, using a custom User-Agent string.
    #   Web.open('http://www.wired.com/', :user_agent => 'the future')
    #
    def Web.open(url,options={})
      user_agent_alias = options.delete(:user_agent_alias)
      proxy = (options.delete(:proxy) || Web.proxy)
      user = options.delete(:user)
      password = options.delete(:password)
      content_length_proc = options.delete(:content_length_proc)
      progress_proc = options.delete(:progress_proc)

      headers = Network::HTTP.headers(options)

      if user_agent_alias
        headers['User-Agent'] = Web.user_agent_aliases[user_agent_alias]
      end

      if proxy[:host]
        headers[:proxy] = Web.proxy_url(proxy)
      end

      if user
        headers[:http_basic_authentication] = [user, password]
      end

      if content_length_proc
        headers[:content_length_proc] = content_length_proc
      end

      if progress_proc
        headers[:progress_proc] = progress_proc
      end

      return Kernel.open(url,headers)
    end

    #
    # Creates a new Mechanize Agent.
    #
    # @param [Hash] options
    #   Additional options.
    #
    # @option options [String] :user_agent_alias
    #   The User-Agent Alias to use.
    #
    # @option options [String] :user_agent
    #   The User-Agent string to use.
    #
    # @option options [Network::HTTP::Proxy, Hash, String] :proxy
    #   (Web.proxy)
    #   Proxy information.
    #
    # @yield [agent]
    #   If a block is given, it will be passed the newly created Mechanize
    #   agent.
    #
    # @yieldparam [WWW::Mechanize] agent
    #   The new Mechanize agent.
    #
    # @return [WWW::Mechanize]
    #   The new Mechanize agent.
    #
    # @example Create a new agent.
    #   Web.agent
    #
    # @example Create a new agent, with a custom User-Agent alias.
    #   Web.agent(:user_agent_alias => 'Linux Mozilla')
    #
    # @example Create a new agent, with a custom User-Agent string.
    #   Web.agent(:user_agent => 'wooden pants')
    #
    # @see http://mechanize.rubyforge.org/mechanize/WWW/Mechanize.html
    #
    def Web.agent(options={},&block)
      agent = WWW::Mechanize.new

      if options[:user_agent_alias]
        agent.user_agent_alias = options[:user_agent_alias]
      elsif options[:user_agent]
        agent.user_agent = options[:user_agent]
      elsif Web.user_agent
        agent.user_agent = Web.user_agent
      end

      proxy = if options[:proxy].kind_of?(Hash)
                options[:proxy]
              elsif options[:proxy].kind_of?(String)
                Network::HTTP::Proxy.parse(options[:proxy])
              elsif options[:proxy].nil?
                Web.proxy
              else
                raise(RuntimeError,"the given :proxy option is neither a Proxy, Hash or String",caller)
              end

      if proxy[:host]
        agent.set_proxy(proxy[:host],proxy[:port],proxy[:user],proxy[:password])
      end

      block.call(agent) if block
      return agent
    end

    #
    # Creates a Mechanize Page for the contents at a given URL.
    #
    # @param [URI::Generic, String] url
    #   The URL to request.
    #
    # @param [Hash] options
    #   Additional options.
    #
    # @option options [String] :user_agent_alias
    #   The User-Agent Alias to use.
    #
    # @option options [String] :user_agent
    #   The User-Agent string to use.
    #
    # @option options [Network::HTTP::Proxy, Hash] :proxy (Web.proxy)
    #   Proxy information.
    #
    # @yield [page]
    #   If a block is given, it will be passed the page for the requested
    #   URL.
    #
    # @yieldparam [WWW::Mechanize::Page] page
    #   The requested page.
    #
    # @return [WWW::Mechanize::Page]
    #   The requested page.
    #
    # @example
    #   Web.get('http://www.rubyinside.com')
    #   # => WWW::Mechanize::Page
    #
    # @example
    #   Web.get('http://www.rubyinside.com') do |page|
    #     page.search('div.post/h2/a').each do |title|
    #       puts title.inner_text
    #     end
    #   end
    #
    # @see http://mechanize.rubyforge.org/mechanize/WWW/Mechanize/Page.html
    #
    def Web.get(url,options={},&block)
      page = Web.agent(options).get(url)

      block.call(page) if block
      return page
    end

    #
    # Requests the body of the Mechanize Page created from the response
    # of the given URL.
    #
    # @param [URI::Generic, String] url
    #   The URL to request.
    #
    # @param [Hash] options
    #   Additional options.
    #
    # @option options [String] :user_agent_alias
    #   The User-Agent Alias to use.
    #
    # @option options [String] :user_agent
    #   The User-Agent string to use.
    #
    # @option options [Network::HTTP::Proxy, Hash] :proxy (Web.proxy)
    #   Proxy information.
    #
    # @yield [body]
    #   If a block is given, it will be passed the body of the page.
    #
    # @yieldparam [String] body
    #   The requested body of the page.
    #
    # @return [String]
    #   The requested body of the page.
    #
    # @example
    #   Web.get_body('http://www.rubyinside.com') # => String
    #
    # @example
    #   Web.get_body('http://www.rubyinside.com') do |body|
    #     puts body
    #   end
    #
    def Web.get_body(url,options={},&block)
      body = Web.get(url,options).body

      block.call(body) if block
      return body
    end

    #
    # Posts to a given URL and creates a Mechanize Page from the response.
    #
    # @param [URI::Generic, String] url
    #   The URL to request.
    #
    # @param [Hash] options
    #   Additional options.
    #
    # @option options [Hash] :query
    #   Additional query parameters to post with.
    #
    # @option options [String] :user_agent_alias
    #   The User-Agent Alia to use.
    #
    # @option options [String] :user_agent
    #   The User-Agent string to use.
    #
    # @option options [Network::HTTP::Proxy, Hash] :proxy (Web.proxy)
    #   Proxy information.
    #
    # @yield [page]
    #   If a block is given, it will be passed the page for the requested
    #   URL.
    #
    # @yieldparam [WWW::Mechanize::Page] page
    #   The requested page.
    #
    # @return [WWW::Mechanize::Page]
    #   The requested page.
    #
    # @example
    #   Web.post('http://www.rubyinside.com')
    #   # => WWW::Mechanize::Page
    #
    def Web.post(url,options={},&block)
      query = {}
      query.merge!(options[:query]) if options[:query]

      page = Web.agent(options).post(url,query)

      block.call(page) if block
      return page
    end

    #
    # Posts to a given URL and returns the body of the Mechanize Page
    # created from the response.
    #
    # @param [URI::Generic, String] url
    #   The URL to request.
    #
    # @param [Hash] options
    #   Additional options.
    #
    # @option options [Hash] :query
    #   Additional query parameters to post with.
    #
    # @option options [String] :user_agent_alias
    #   The User-Agent Alias to use.
    #
    # @option options [String] :user_agent
    #   The User-Agent string to use.
    #
    # @option options [Network::HTTP::Proxy, Hash] :proxy (Web.proxy)
    #   Proxy information.
    #
    # @yield [body]
    #   If a block is given, it will be passed the body of the page.
    #
    # @yieldparam [WWW::Mechanize::Page] page
    #   The body of the requested page.
    #
    # @return [WWW::Mechanize::Page]
    #   The body of the requested page.
    #
    # @example
    #   Web.post_body('http://www.rubyinside.com')
    #   # => String
    #
    # @example
    #   Web.post_body('http://www.rubyinside.com') do |body|
    #     puts body
    #   end
    #
    def Web.post_body(url,options={},&block)
      body = Web.post(url,options).body

      block.call(body) if block
      return body
    end
  end
end
